Added tails file prime generation and tails file hashing

aritroCoder opened the pull request:
Before:
![image](https://github.com/hyperledger/anoncreds-spec/assets/92646038/18d4fc6e-9662-484a-94b0-61752ba47ce2)

After:
![image](https://github.com/hyperledger/anoncreds-spec/assets/92646038/b7124c3f-4985-4f6f-a2df-e6c4a9ec98b6)
![image](https://github.com/hyperledger/anoncreds-spec/assets/92646038/87186051-93c8-494f-91f8-06112b08d200)

andrewwhitehead commented:
The entries in the tails file aren't actually primes, they are G2 curve points. The process for generating them looks correct here. Essentially, it is the case that `tail[index] = g_dash * (gamma ** index)`, where `gamma` is a scalar field element. I'm not sure I would describe `g_dash` as 'the' generator of the G2 group as it seems to be picked randomly for each registry.

aritroCoder commented:
> The entries in the tails file aren't actually primes, they are G2 curve points. The process for generating them looks correct here. Essentially, it is the case that `tail[index] = g_dash * (gamma ** index)`, where `gamma` is a scalar field element. I'm not sure I would describe `g_dash` as 'the' generator of the G2 group as it seems to be picked randomly for each registry. Every curve point is technically a generator but there is also a standard one that is used elsewhere.

![image](https://github.com/hyperledger/anoncreds-spec/assets/92646038/3e20dd78-75ac-425a-898e-a53d6c32e06c)
This text is from another part of the same specification where they call G2 as the generator

aritroCoder commented:
I have updated the first message with the new updated screenshot of edited spec. Also, I have changed the mention of  `primes` with `points in curve G2` in that section (I guess it needs a separate issue to track down and replace all such instances in case this is present elsewhere too)

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -444,30 +444,42 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
+an array of 1000 `G2` curve points. Each credential issued using the [[ref: Revocation Registry]] is
 given its own index (1 to the capacity of the [[ref: Revocation Registry]]) into the array,
-the index of the prime for that credential. The contents of the [[ref;
+the index of the point for that credential. The contents of the [[ref;
```
Some cleanup of text around what you changed. I’ll add a few of these.

Fix the “;” in the “ref:” at the end of this line.

aritroCoder replied:
fixed

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -444,30 +444,42 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
 
-::: todo
-To Do: Document the process for generating the primes.
+The process of generating the primes that populate the [[ref: TAILS_FILE]] are `tail[index] = g_dash * (gamma ** index)`
```
Should the reference to “primes” here be changed as well?

BTW — are “points on the curve G2” prime numbers?

Question: Should we define in the reference section what G1 and G2 are?  Is that needed?  I don’t know — I just sort of gloss over them.  And those that know anything about this stuff probably don’t need an explanation?

aritroCoder replied:
Yes, I missed that line, i'll change it now

I dont think "points on curve G2" are prime. I think there might be some discrepancy between the specification and actual implementation (like it was initially thought to be implemented as prime numbers but later while coding they found out using points could be easier).

G1 and G2 are defined a few sections before in the spec. I am attaching a screenshot of the portion of the spec:

![image](https://github.com/hyperledger/anoncreds-spec/assets/92646038/b3bfe2d3-f097-4b37-90cd-1540390fe424)

But we can add the definitions of G1 and G2 for this context here, that can make this section easier for newcomers

swcurran replied:
Note that definitions can have synonyms, so a single definition would be enough for references to the terms “G1”, “G2” at least, perhaps also “pairing friendly elliptical curve”.

aritroCoder replied:
yeah, so then I guess it's fine as it is

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -444,30 +444,42 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
-To Do: Document the process for generating the primes.
+The process of generating the primes that populate the [[ref: TAILS_FILE]] are `tail[index] = g_dash * (gamma ** index)`
+
+::: note
```
Should this be in a note section?  I’m not sure.

Should the actual math expressions in the bullet points be in Latex (as well or instead)?  Does that make it clearer/more useful?

aritroCoder replied:
I think here we are dealing with how variables are expressed in the actual code rather than in latex (for example we write `g_dash` in the spec instead of writing it in latex). So using latex won't be much useful in my opinion, as it might instead create confusion to the reader.

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -444,30 +444,42 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
+- Repeat for all the attributes.
+- Close the file buffer.
+
+Relevant links: [Anoncreds-rs repository](https://github.com/hyperledger/anoncreds-rs/blob/9c915bb77bc4e033cc6d28d45e330ee5bda26211/src/services/tails.rs#LL148C1-L148C37), [Ursa repository](https://github.com/hyperledger-archives/ursa/blob/c29fdaa96bbe9ce3ea2beb4d5fbe98ed7c96f867/libursa/src/cl/mod.rs#L514)
```
Please switch the reference to the Ursa repository in favour of the new https://github.com/hyperledger/anoncreds-clsignatures-rs repository.

aritroCoder replied:
done

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -444,30 +444,42 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
 
-Once generated, the array of primes is static, regardless of credential issuance
+Once generated, the array of points is static, regardless of credential issuance
 or revocation events. Once generated, the SHA256 (TO BE VERIFIED) hash of the
```
I think the “(TO BE VERIFIED)” can be removed.

swcurran commented:
Looks great — nice work.  A few minor cleanups mostly in the text that existed before you started.

Not sure about the “To Do” right below your PR.  Is there something you can add about for the format of the file (e.g. how to extract the points from it), or should it be removed?

aritroCoder commented on spec/data_flow_setup.md:
```diff
@@ -444,30 +444,42 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
 
-The process of generating the primes that populate the [[ref: TAILS_FILE]] is as
-follows:
+The process for hashing the [[ref: TAILS_FILE]] is as follows:
```
yeah did it

aritroCoder commented:
> Looks great — nice work. A few minor cleanups mostly in the text that existed before you started.
> 
> Not sure about the “To Do” right below your PR. Is there something you can add about for the format of the file (e.g. how to extract the points from it), or should it be removed?

for the `TO DO` where we have to define the format of the tails file, I am guessing we have to write what all data is written in the file, and in which order.

aritroCoder commented:
> The values are elliptic curve points and not primes encoded. I need to look deeper as to how those points are encoded

Yeah, that's the reason why the are encoded to the octet stream, which makes it difficult to create a human understandable format

andrewwhitehead commented:
The points are encoded as uncompressed big-endian coordinates: `{x.0, x.1, y.0, y.1}` for 32*4 = 112 bytes, for the two Fp2 components.

aritroCoder commented:
I have updated the first message with the new updated screenshot of edited spec after adding the tails file format

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -486,7 +486,7 @@ section](#revocation-registry-definition-object-generation)). Typically, the arr
 the `tailsLocation` input parameter provided by the [[ref: issuer]].
 
-The format of a [[ref: TAILS_FILE]] is as follows(binary points are shown in readable format):
+The format of a [[ref: TAILS_FILE]] is as follows(bytes representing points are shown in readable format):
```
I’m not exactly clear on the format and have a couple of questions that I think might help:

- Likely my ignornance, but I’m not familiar with the “u” notation. I’m guessing it is 2 bytes at the beginning of the file?
- Is the “version:” literal?
- How long (number of bytes) are each of the points?

aritroCoder replied:
- `0u8` is the number 0 as an unsigned 8-bit integer, similiarly for `2u8`
- Version is specifically the version of the file(which is there in the rust implementation [here](https://github.com/hyperledger/anoncreds-rs/blob/9c915bb77bc4e033cc6d28d45e330ee5bda26211/src/services/tails.rs#L153C16-L153C16)) 
- A point is a collection of two integers, so it should be 2*4 bytes(for 2D point), but since `ECP2` implemented points as 3D, so it would be 3*4=12 bytes long

swcurran replied:
Got it.  I would suggest then making the file format just bullet points (vs. fenced) and describe what you have above, something like:

* The version of the tails file, as a 2 unsigned bytes. The current version uses the value is `0u8 2u8`.
* A list of the points, one per credential in the Revocation Registry. Each a collection of two integers (<what is in the PR now> implemented as points in 3 dimensions. Each point is 12 bytes long.

The total size of a Tails File is `2+12*Size of the Revocation Registry`.

aritroCoder replied:
Did it, and updated the screenshots in the first message

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -444,40 +444,54 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
+- Calculate `pow` by doing modular exponentiation of revocation private key(`gamma`) with the finite field element previously calculated.
+- Multiply `pow` by `g_dash`, which is the generator of elliptic curve group `G2`, and this should be the required point on the curve.
+- Convert this tail point to an array of bytes(`u8`), and put them into the file as a slice buffer.
+- Repeat for all the attributes from index $1$ to $L$, by calculating $([\gamma], [\gamma^2], [\gamma^3], ...[\gamma^L], [\gamma^{L+2}], [\gamma^{L+3}], ..., [\gamma^{2L}])$. Omit $[\gamma^{L+1}]$ as it can be used to generate fake witness for non-revocation.
```
To be discussed -- I think this is not right. From what I understand:

- 2 bytes for version
- 1...$L are points associated with credentials 1...L
- $L+1 is a duplicate of 1 (filler)
- $L+2...$2L+1 are points associated with credentials 1...L

Hopefully we can clarify this on the July 31, 2023 AnonCreds call.

swcurran replied:
In the sequence part `$([\gamma], [\gamma^2], [\gamma^3], ...[\gamma^L], [\gamma^{L+2}], [\gamma^{L+3}], ..., [\gamma^{2L}])$` add in `[\gamma]` where L+1 has been left out.

Remove the "Omit..." sentence and replace it with `Note that Instead of inserting $[\gamma^{L+1}]$ in the sequence, insert the value $[\gamma]$ (the first value in the sequence) in its place, and then continue with $[\gamma^{L+2}]$ and on to $[\gamma^{2L}]$. $[\gamma^{L+1}]$ is not used by holders generating the [[def: Non-Revocation Proof]] and a dummy value is inserted in its place.`

swcurran replied:
@andrewwhitehead -- can you confirm my phrasing?

swcurran replied:
@aritroCoder -- confirmed with Andrew -- `It's a copy of index 0, which isn't used otherwise. g_dash*gamma^0 which is just g_dash`

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -444,40 +444,54 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
-:::
+The process for hashing the [[ref: TAILS_FILE]] is as follows:
+
+- Use a hasher with any hashing algorithm(`SHA256` in [anoncreds-rs implementation](https://github.com/hyperledger/anoncreds-rs/blob/9c915bb77bc4e033cc6d28d45e330ee5bda26211/src/services/tails.rs#LL148C1-L148C37)).
```
I think we need to explicitly declare this is `SHA256` so that others can implement AnonCreds and be compatible with the AnonCreds Rust implementation.

swcurran replied:
Change -- "Use a hasher with any hashing algorithm(`SHA256` in" to "Use the `SHA256` hashing algorithm"

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -444,40 +444,54 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
+- First two bytes are version number(currently `0u8 2u8`)
+- A list of the points, one per credential in the Revocation Registry. Each point is a collection of three integers implemented as points in 3 dimensions as per `ECP2`. Each point is 3x4 = 12 bytes long.
+
+Thus the total size of a Tails File is 2 + 12*`Size of the Revocation Registry`.
```
I think this should 2+ 12*Size of the Revocation Registry+6 (the L+1 entry).

swcurran replied:
Each point is 256 bits or 32 bytes. Each coordinate is two field elements, so a total of 4 elements, and 128 bytes per point.

swcurran replied:
@andrewwhitehead -- can you please confirm the correct size of the file?

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -444,40 +444,54 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
 given its own index (1 to the capacity of the [[ref: Revocation Registry]]) into the array,
-the index of the prime for that credential. The contents of the [[ref;
+the index of the point for that credential. The contents of the [[ref:
 TAILS_FILE]] is needed by the [[ref: issuer]] to publish the current state of
```
This has changed recently in the implementation. Please remove "by the [[ref: issuer]] to publish the current state of revocations within the [[ref: Revocation Registry]] and"

swcurran commented on spec/data_flow_setup.md:
```diff
@@ -444,40 +444,54 @@ The identifier for the [[ref: Revocation Registry]] is dependent on where the
+
+Once generated, the array of points is static, regardless of credential issuance
+or revocation events. Once generated, the SHA256 hash of the
+array of points is calculated and returned to be inserted into the `tailsHash`
```
Change this:  

- Once generated, the SHA256 hash of the array of points is calculated and returned to be inserted into the `tailsHash`

to this:

-  The SHA256 hash of the array of points is returned to be inserted into the `tailsHash`
